File: weather-demo-app-apple/Weather/UI/Current/CurrentWeatherViewModel.swift
//
//  CurrentWeatherViewModel.swift
//  Weather
//
//  Created by Onur on 4/11/26.
//

import Foundation
import Combine

@MainActor
final class CurrentWeatherViewModel: ObservableObject {
    @Published private(set) var state: CurrentWeatherViewState = .idle
    @Published var city: String = ""

    private let repository: WeatherDataSource
    private var hasLoadedInitialCity = false
    private var lastLoadedCity: String?
    private var searchTask: Task<Void, Never>?

    init(repository: WeatherDataSource) {
        self.repository = repository
    }

    func onCityChanged(_ value: String) {
        city = value
    }

    func loadLastCity() async {
        guard !hasLoadedInitialCity else { return }
        hasLoadedInitialCity = true

        let lastCity = repository.getLastSearchedCity().orEmpty
        guard !lastCity.isEmpty else { return }

        city = lastCity
        await search(city: lastCity)
    }

    func search(city: String, force: Bool = false) async {
        let targetCity = city.trimmingCharacters(in: .whitespacesAndNewlines)

        if !force,
           lastLoadedCity == targetCity,
           case .success = state {
            return
        }

        searchTask?.cancel()

        searchTask = Task {
            state = .loading

            do {
                let result = try await repository.getCurrentWeather(city: targetCity)
                self.city = result.city
                self.lastLoadedCity = result.city
                self.state = .success(result)
            } catch let error as RepositoryError {
                self.state = .error(error.message)
            } catch {
                self.state = .error(AppConstants.genericErrorMessage)
            }
        }

        await searchTask?.value
    }

    func retry() async {
        await search(city: city, force: true)
    }
}

File: weather-demo-app-android/app/src/main/java/com/demo/weather/ui/current/CurrentWeatherViewModel.kt
package com.demo.weather.ui.current

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.demo.weather.app.AppConstants
import com.demo.weather.data.WeatherDataSource
import com.demo.weather.domain.model.RepositoryError
import com.demo.weather.domain.model.toMessage
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class CurrentWeatherViewModel(
    private val repository: WeatherDataSource
) : ViewModel() {

    private val _state = MutableStateFlow<CurrentWeatherViewState>(CurrentWeatherViewState.Idle)
    val state: StateFlow<CurrentWeatherViewState> = _state.asStateFlow()

    private val _city = MutableStateFlow("")
    val city: StateFlow<String> = _city.asStateFlow()

    private var hasLoadedInitialCity = false
    private var lastLoadedCity: String? = null
    private var searchJob: Job? = null

    fun onCityChanged(value: String) {
        _city.value = value
    }

    fun loadLastCity() {
        if (hasLoadedInitialCity) return
        hasLoadedInitialCity = true

        val lastCity = repository.getLastSearchedCity().orEmpty()
        if (lastCity.isEmpty()) return

        _city.value = lastCity
        search(city = lastCity)
    }

    fun search(city: String, force: Boolean = false) {
        val targetCity = city.trim()

        if (!force &&
            lastLoadedCity == targetCity &&
            _state.value is CurrentWeatherViewState.Success
        ) {
            return
        }

        searchJob?.cancel()

        searchJob = viewModelScope.launch {
            _state.value = CurrentWeatherViewState.Loading

            try {
                val result = repository.getCurrentWeather(city = targetCity)
                _city.value = result.city
                lastLoadedCity = result.city
                _state.value = CurrentWeatherViewState.Success(result)
            } catch (error: RepositoryError) {
                _state.value = CurrentWeatherViewState.Error(error.toMessage())
            } catch (_: Throwable) {
                _state.value = CurrentWeatherViewState.Error(AppConstants.GENERIC_ERROR_MESSAGE)
            }
        }
    }

    fun retry() {
        search(city = _city.value, force = true)
    }
}